README.md
# SKCC SKIMMER WEBAPP

This is a web wrapper to [K7MJG's SKCC skimmmer][1] python application.

It parses the output of a out-of-the box configuration of his application and displays
them in a web page. This webpage is mobile friendly, too.

Behind the scenes, the web app will kick off a process for K7MJG's SKCC skimmer
application. You won't normally see the output of this process. The web app will
translate the output that you would normally see and places it in a convenient
website that you can browse **from inside your network**. 

**It's not recommended that you expose this website to the outside world. You 
do so at your own risk!!!**

**Main web app User Interface:**
<p align="center">
<img src="https://raw.githubusercontent.com/cwhelchel/skimmer_webapp/master/docs/img/screenshot1.png?raw=true" width="600" height="400" alt="Skimmer Screenshot" />
</p>

**Info from SKED page:**
<p align="center">
<img src="https://raw.githubusercontent.com/cwhelchel/skimmer_webapp/master/docs/img/screenshot2.png?raw=true" width="600" height="50" alt="SKED Screenshot" />
</p>

Note: the green text is what you get from a QSO from this member.

**Info from a RBN SPOT:**
<p align="center">
<img src="https://raw.githubusercontent.com/cwhelchel/skimmer_webapp/master/docs/img/screenshot3.png?raw=true" width="600" height="30" alt="SKED Screenshot" />
</p>

Note: This is highlighted blue because its the first time the skimmer saw this member
and you need them for goals or targets. They also may need you for their target hence the "/ they: C".

## Configuring SKCC SKIMMER

Once you've downloaded either the release or the source directly, you must configure
the SKCC SKIMMER in order for it to work correctly. If you have used [K7MJG's SKCC skimmmer][1] before then you have already done this and the rest should be easy.
If not there's only a few lines to configure.

If you have already run K7MJG's skimmer before then you just need to copy and paste  your existing ```skcc_skimmer.cfg``` over the included file of the same name. If you have never run the skimmer before, open the included file and update these three things: 

* Your callsign: `MY_CALLSIGN`
* Your QTHs Maidenhead GridSquare: `MY_GRIDSQUARE`
* the path to your master adi file (skcclogger): `ADI_FILE`

For example, here's what my ```skcc_skimmer.cfg``` looks like in the file (this 
is on Windows 10):

```python
    MY_CALLSIGN    = 'KQ4DAP'           
    MY_GRIDSQUARE  = 'EM82dl'           
    SPOTTER_RADIUS = 750          

    ADI_FILE       = r'C:\\SKCCLogger\\Logs\\skcc.adi'
```

You could also take the time to change your goals, targets, and bands. Refer to
the [skcc_skimmer configuration][2] for more information.

To configure web-app specific options open ```skimmerwebapp.cfg```.

## Running the Webapp - Windows

This section is for Windows users:

1) Download the release package
2) Extract to a location you will be able to find again
3) Update the included skcc_skimmer.cfg file inside the package by either:
   - copying an existing skcc_skimmer.cfg into here
   - open the file and update the needed settings
4) Execute the ```run_webapp.bat``` file.
5) Point your browser to http://127.0.0.1:5000 to see the website

You should see a command prompt appear with server status information and some other
output. This is the skimmer running and printing. It prints out a lot. If needed
this can be disabled, look in ```skimmwerwebapp.cfg```

## Running the Webapp - Others

This section is for non-Windows users and other cool people.

Run directly from source using Python. It's really not that hard, and the 
process is explained in the next sections:

### Install Dependencies

You need Python 3, Flask, and Flask_sock.

After installing Python, at the command prompt run these:

    $ python -m pip install flask
    $ python -m pip install flask_sock

Or, optionally

    $ python -m pip install -r requirements.txt    
    #  ^^^ will include pyinstaller which is needed for exe releases

### Running the WEB APP    

Once this file is updated and saved, you are ready to run the application at the command prompt, like so:

    $ python flask_app.py

Browse to the location it outputs (http://127.0.0.1:5000) and enjoy.

## Running with Docker

Coming soon. Hopefully!

73,

\- Cainan KQ4DAP


[1]: https://github.com/k7mjg/skcc_skimmer
[2]: https://www.k7mjg.com/#id_Configuration

docs/CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

### Added

## [0.1.1] - 2023-03-06

### Fixed

- Windows release package should work without having to install Python
- Linux release should work but does require Python


## [0.1.0] - 2023-02-21

### Added

- Initial release of functional (hopefully) SKCC Skimmer Webapp
- Documentation

[0.1.0]: https://github.com/cwhelchel/skimmer_webapp/commits/v0.1.0
[0.1.1]: https://github.com/cwhelchel/skimmer_webapp/commits/v0.1.1

app.py
import subprocess
import io
import re
from time import sleep
from skimmer import cSkimmer
from skimmer import Parser

# spawn a subprocess that redirects the main skcc_skimmer output from stdout
# to a file (the skimmer displays info only at the console)
def start_skimmer():
    x = cSkimmer()
    x.run()
    return x

data = "1906Z W2NR   ( 1050 Sx6  Randy      NH); YOU need them for K3Y/1 (20m); STATUS: K3Y/1 working down the bands 14.117"
#"2357Z+VE3KZ  (22804 Sx5  Bob        ON) on   7059.0 by K9LC(734mi, 37dB); YOU need them for BRAG,C,P(new +22804)"
# "0009Z WB4HNL (22010      Rick       FL) on   7052.0 by WE9V(735mi, 17dB); YOU need them for BRAG,C,WAS,P(new +22010); THEY need you for C"
# "0013Z W0GV   ( 7889 S    Gerry      CO) on   7055.5 by K9LC(734mi, 19dB); YOU need them for BRAG,C,WAS-S,P(new +7889)"
# 
# 0023Z N4DT   (16533      Danny      SC) on   7055.9 by W8WWV(637mi, 13dB); YOU need them for BRAG,C,WAS; THEY need you for C
# 0023Z N4DT   (16533      Danny      SC) on   7055.9 by W8WTS(631mi, 9dB); YOU need them for BRAG,C,WAS; THEY need you for C"    

if __name__ == '__main__':
    print("test")

    spot = Parser.parse_spot(data)

    print(spot)

    # if m:
    #     print(m.groups())
    #     print("ih")
    
    # proc = start_skimmer()

    # while True:
    #     print(proc.read())
    #     sleep(500)

skimmer.py
import os, sys
import subprocess
import threading
from skimmer_parser import SkimmerParser
from datetime import datetime, timezone, timedelta
from skimmer_spot import Spot, SpotType


class SkccStatus:
    __config = {}
    __lines = []
    
    state = 'starting'

    def __str__(self) -> str:
        return '\r\n'.join(self.__lines)

    def add_cfg(self, cfg: str, data: str):
        self.__config[cfg] = data

    def add(self, line : str):
        self.__lines.append(line)

    def get_lines(self) -> list:
        return [self.state] + self.__lines


class cSkimmer:
    """The main class for parsing output from skcc_skimmer.py

    This class will spawn a python subprocess (of skcc_skimmer.py) with piped
    stdout which is then read forever. Call run() then read() <= which blocks 
    forever

    :param cfg: dictionary of configured values. Usually the ```app.config``` 
    object from Flask
    """

    def __init__(self, cfg : dict):
        if sys.platform == "win32":
            self.__cmd = ["skcc_skimmer.exe"]
        else:
            self.__cmd = ["python3", "skccskimmer/skcc_skimmer.py"]
        self.__status = SkccStatus()
        self.__spots = []
        self.__sked_spots = []
        self.__new_spot = False
        self.__new_skeds = False
        self.__parsing_skeds = True
        self.__is_running = False
        self.__config = cfg

    def run(self):
        """Create the python subprocess and run the main skimmer application.
        
        This should be done first.
        """

        if not self.__is_running:
            os.environ['PYTHONIOENCODING'] = 'utf-8'

            print('running cmd: ' + ' '.join(self.__cmd))
            print('in curr dir: ' + os.getcwd())

            self.__proc = subprocess.Popen(self.__cmd, 
                encoding='utf-8', 
                stdout=subprocess.PIPE, 
                stderr=subprocess.PIPE,
                stdin=subprocess.PIPE)

            self.__is_running = True

            # kick off parsing thread
            self.__thread = threading.Thread(target=self.read)
            self.__thread.start()
            print("read thread started")


    def stop(self):
        """Stops the running skcc skimmer process
        
        """
        if (self.__proc != None and self.__is_running):
            print('killing process...')
            self.__proc.terminate()

            print('joining read thread...')
            self.__thread.join()

            self.__is_running = False
            self.__status.state = 'stopped'


    def read(self):
        """Reads and parses the skimmer subprocess' output. 
        
        This never returns.
        """

        if (self.__proc == None):
            return ""

        self.__status.state = 'starting'

        str = ""
        for line in iter(self.__proc.stdout.readline, ''):
            str = line.strip()
            self.__parse(str)

            if "SHOW_SKIMMER_OUTPUT" in self.__config and \
                self.__config['SHOW_SKIMMER_OUTPUT']:
                print(str)

        # we never get here
        return str


    def get_spots(self) -> tuple[bool, list]:
        """Returns the a tuple with the lists of spots and a flag indicating there are new spots
        
        Calling this method sets the above flag to False
        """
        result = (self.__new_spot, self.__spots)
        self.__new_spot = False
        return result


    def get_skeds(self) -> tuple[bool, list]:
        """Returns the a tuple with the lists of skeds and a flag indicating there are new skeds
        
        Calling this method sets the above flag to False. Also if this method is
        called while new skeds are being parsed. It returns the old skeds.
        """
        if (self.__parsing_skeds):
            return (False, self.__sked_spots)

        result = (self.__new_skeds, self.__sked_spots)
        self.__new_skeds = False
        return result


    def get_status(self) -> SkccStatus:
        ''' Gets the skimmer status object '''
        return self.__status


    def clear_spots(self):
        """ Clears the list of current spots (not skeds) from the object
        """
        self.__spots.clear()

        # force refresh of frontend spots list
        self.__new_spot = True


    def force_refresh(self):
        '''Sets the flags to indicate that there are new spots to see.'''
        self.__new_skeds = True
        self.__new_spot = True


    def __parse(self, line : str):
        if (self.__status.state == 'starting'):
            # if we're starting lets grab all the output for later 
            self.__parse_cfg_line(line, "GOALS:")
            self.__parse_cfg_line(line, "TARGETS:")
            self.__parse_cfg_line(line, "BANDS:")
            self.__status.add(line)
            if line.startswith("Running..."): 
                self.__status.state = 'running'

        if (self.__status.state == 'running'):
            # were running so lets parse all the juicy spots
            if (line == "=========== SKCC Sked Page ============"):
                print('new skeds incoming...')
                self.__parsing_skeds = True
                self.__sked_spots.clear()
            elif (line == "======================================="):
                print('skeds done.')
                self.__parsing_skeds = False
                self.__new_skeds = True
            else:
                self.__parse_line(line)

            self.__check_spots()


    def __parse_cfg_line(self, line : str, prefix : str):
        if line.startswith(prefix):
            self.__status.add_cfg(prefix, line)


    def __parse_line(self, line : str):
        spot = SkimmerParser.parse_spot(line)

        if (spot == None): 
            return

        #print(spot)

        if (self.__parsing_skeds):
            spot.kind = SpotType.SKED
            spot.spotter = "sked"
            self.__sked_spots.append(spot)
        else:
            spot.kind = SpotType.SPOT
            self.__new_spot = True
            self.__spots.append(spot)


    def __check_spots(self):
        '''Loop thru the spots list and remove any older than a certain time'''
        if "SPOT_EXPIRE_TIME" in self.__config:
            timeout = self.__config["SPOT_EXPIRE_TIME"]
        else:
            timeout = 3600

        for spot in self.__spots:
            t = datetime.now(tz=timezone.utc)
            # print(f'spot time {spot.utc_time} == current {t}')
            delta = t - spot.utc_time

            # if its been an hour remove the spot from the list
            if delta >= timedelta(seconds=timeout):
                print("removing old spot")
                self.__spots.remove(spot)


flask_app.py
from datetime import timedelta
from time import sleep
from flask import Flask, render_template
from flask import jsonify, request, session
from flask_sock import Sock
from skimmer import cSkimmer

WEB_APP_VER = "0.1.0"

app = Flask(__name__)

# setup config for sessions
app.config['SESSION_PERMANENT'] = True
app.config['SESSION_TYPE'] = 'filesystem'
app.config['PERMANENT_SESSION_LIFETIME'] = timedelta(hours=48)
app.secret_key = 'WE_SHOULD_PROBABLY_CHANGE_THIS'

sock = Sock(app)
app.config.from_pyfile('skimmerwebapp.cfg')
skimmer = cSkimmer(app.config)


@app.before_first_request
def start_skimmer():
    # make session permanent
    session.permanent = True

    # kick off main skcc skimmer script
    skimmer.run()


@app.route("/")
def index():
    """Getting this route will make the skimmer immediately resend the spots
    and skeds.
    """
    skimmer.force_refresh()
    return render_template('index.html')


@app.route("/restart/", methods=['POST'])
def restart():
    """Posting to this route will stop the skimmer then restart it
    """
    skimmer.stop()
    skimmer.force_refresh()
    skimmer.run()
    return render_template('index.html')    


@app.route('/getstatus')
def get_status():
    """Gets the status of the SKCC Skimmer background application (at startup).
    
    This method gets all the text output lines of the skimmer during it's initial
    startup period. Afterwards it serves no function.
    """
    s = skimmer.get_status()
    j = s.get_lines()
    return jsonify(j)


@app.route('/clearspots')
def clear_spots():
    """Allow user to clear the current list of RBN spots.
    
    Returns "nothing" str
    """
    skimmer.clear_spots()
    return "nothing"


@app.route('/save/', methods=['POST'])
def save():
    """Allows a user to save a SPOT line from the UI to the current session.
    
    POSTing with ?clear=1 will pop the current list of stored_spots from the 
    session.
    """
    data = request.get_json()
    #print(data)
    if not session.get('stored_spots'):
        session['stored_spots'] = []

    session['stored_spots'].append(data)
    session.modified = True

    do_clear = request.args.get('clear', default=0, type=int)

    if (do_clear == 1):
        session.pop('stored_spots')

    return jsonify(success=True)


@app.route('/load', methods=['GET'])
def load():
    """Returns the session's stored_spots array as a JSON object.
    
    Use this on refresh to get the current stored spots.
    Returns {Success}
    """
    if not session.get('stored_spots'):
        return jsonify(success=False, message='no stored spots')

    print('load' + str(session['stored_spots']))
    return jsonify(session['stored_spots'])


@app.route('/version', methods=['GET'])
def version():
    """Returns the applications version information. 

    """

    return jsonify(success=True, version=WEB_APP_VER)

@sock.route('/getskedsws')
def get_skeds_ws(ws):
    """Provide a websocket connection that continually sends the current list of
    skeds.
    
    This loops forever and delays based on the app.config 'SKED_TIME' 
    """
    while True:
        sleep(app.config['SKED_TIME'] / 1000)
        (new_skeds, skeds) = skimmer.get_skeds()

        x = jsonify(skeds)
        ws.send(x.data.decode('utf-8'))


@sock.route('/getspotsws')
def get_spots_ws(ws):
    """Provide a websocket connection that continually sends the current list of
    RBN spots of SKCC members.
    
    This loops forever and delays based on the app.config 'SPOT_TIME' 
    """
    while True:
        sleep(app.config['SPOT_TIME'] / 1000)
        (new_spots, spots) = skimmer.get_spots()

        x = jsonify(spots)
        ws.send(x.data.decode('utf-8'))


# when we run out of a bundled exe this is what starts off the flask application
if __name__ == '__main__':

    # run the flask server
    if (app.config['BIND_LOCALHOST_ONLY']): 
        app.run()
    else:
        app.run(host="0.0.0.0")

skimmer_spot.py
from datetime import datetime, timezone
from dataclasses import dataclass

class SpotType:
    SKED = 1
    SPOT = 2
    UNKNOWN = 3

@dataclass
class Spot:
    '''This is the data object for a spot line.
    
    Objects of this type will be sent to the front end for display.'''

    time : str
    call : str
    kind : SpotType
    flag : bool = False
    freq : str = ""
    spotter : str = ""
    name : str = ""
    spc : str = ""
    num : str = ""
    you_need: str = ""
    they_need: str = ""
    sked_stat: str = ""
    utc_time : datetime = datetime.min

    def __post_init__(self):
        self.time = self.time[:2] + ':' + self.time[2:].lower()
        hour = self.time[:2]
        minute = self.time[3:5]
        self.utc_time = datetime.utcnow()
        self.utc_time = datetime(self.utc_time.year, \
            self.utc_time.month, \
            self.utc_time.day, \
            int(hour), \
            int(minute), \
            0, 0, tzinfo=timezone.utc)


skimmer_parser.py
import re
from skimmer_spot import Spot, SpotType

class SkimmerParser():
    '''This class handles the parsing of output from the skimmer.
    
    The main entry point parse_spot is a static method called like so
    
    `SkimmerParser.parse_spot(line)`
    
    It is used to convert an output line into the `Spot` type the webapp uses.
    '''

    # group 1 = zulu time, 2 = sked/spot flag, 3 = callsign
    spot_hdr = re.compile("([0-9]{4}Z)(\+| )([A-Z0-9\/]*)")

    # group 1 = frequency, 2 = spotter info
    freq = re.compile("on\s+([\d\.\d+]+)\sby\s([\w\-\/\\]+\([\d\w,\s]+\))")

    # group 1 = skcc num, 2 = fname, 3 = SPC
    skcc = re.compile("\(\s*(\d+\s(?:\w|\s)+)\s+([\w\-\']+)\s+([0-9A-Za-z]+)\)")


    @staticmethod
    def parse_spot(line : str) -> Spot:
        '''Parses a line from the skimmer and returns a Spot object.

        May return None if the spot header is invalid. '''
        spot = SkimmerParser._parse_spot_header(line)
        
        if (spot):
            spot = SkimmerParser._parse_spot_body(line, spot)

        return spot


    @staticmethod
    def _parse_spot_header(line : str) -> Spot:

        # sometimes the line has a bell char \x07 at the front (from the main skimmer)
        # we need to remove it (and other control chars) (this isn't all of them)
        line = re.sub(r'[\x00-\x1f]', '', line)

        match = SkimmerParser.spot_hdr.match(line)

        if match:
            x = match.group(1, 2, 3)

            # NOTE: for skeds, the flag indicator is for status updates to the 
            # sked page. for spots its to indicate that that call-sign is needed
            # for a goal.

            flag = True if x[1] == '+' else False

            spot = Spot(time = x[0], \
                        call = x[2], \
                        kind = SpotType.UNKNOWN,\
                        flag = flag)
            return spot
        
        #else: 
            #print("error spot header didn't match!")
        
        return None


    @staticmethod 
    def _parse_spot_body(line : str, spot : Spot) -> Spot:
        x = line.split(';')

        if len(x) <= 0:
            return spot
        
        # parse first split str (freq, name, num, etc.)
        if y := SkimmerParser.freq.search(x[0]):
            (freq, spotter) = y.group(1,2)
            spot.freq = freq.strip()
            spot.spotter = spotter.strip()

        if z := SkimmerParser.skcc.search(x[0]):
            (num, name, spc) = z.group(1,2,3)
            spot.num = num.strip()
            spot.name = name.strip()
            spot.spc = spc.strip()

        # pull out the "needs" and sked status if its there (it may be in idx 2)
        temp = x[1].strip() if len(x) > 1 else ""
        spot.you_need  = x[1][18:] if len(x) > 1 else ""
        spot.they_need = x[2].strip() if len(x) > 2 else ""
        spot.sked_stat = x[3].strip() if len(x) > 3 else ""

        if spot.they_need.startswith("STATUS:"):
            spot.sked_stat = spot.they_need
            spot.they_need = ""
        elif spot.they_need.startswith("THEY"):
            spot.they_need = spot.they_need[17:].strip()

        # if a sked dude is spotted then this gets inserted right after the 
        # callsign and skcc info 
        if temp.startswith("Last spotted"):
            t = re.search("([\d\.\d+]+)", spot.you_need)
            if t:
                spot.freq = t.group(1)
            spot.you_need = spot.they_need[18:]
            spot.they_need = spot.sked_stat[17:] if spot.sked_stat.startswith("THEY need") else ""
            spot.sked_stat = x[4].strip() if len(x) > 4 else spot.sked_stat

        return spot


static/main.js
let at_bottom = false;
var skeds = {};
var spots = {};

/*
*
* --- Local Functions  ---
*
*/


/*
* convert the a JSON spot object into HTML
* 
* param: spot - json object returned from flask app
* returns: string with html representation of the spot param
*/
function getSpotHtml(spot) {
    var out = '';
    var spot_class = "spotline";
    if (spot.flag) {
        spot_class += " newhighlight";
    }

    out += '<div class="' + spot_class + '">';
    out += '<span class="zulu">' + spot.time + '</span>' +
        '<span class="callsign"> ' + spot.call + ' </span>' +
        '<span class="freq">' + spot.freq + '</span>' +
        '<span class="spotter"> ' + spot.spotter + ' </span>' +
        '<span class="skcc_info"> (' + spot.num + ' ' + spot.name + ' ' + spot.spc + ') </span>' +
        '';
    out += '<span class="youneed fw-lighter"> ' + spot.you_need + '</span>';
    if (spot.they_need !== "") {
        out += '<span class="theyneed fw-lighter"> / they: ' + spot.they_need + '</span>';
    }
    out += '<span class="sked_status"> ' + spot.sked_stat + '</span>';
    out += "<br>";
    out += '</div>';
    return out;
}

/* 
*  Write out the HTML for a the spot array (arr) to the 
*  element at (id).
*/
function writeSpotLines(arr, id) {
    var out = '';
    var i;
    
    for (i = 0; i < arr.length; i++) {
        out += getSpotHtml(arr[i]);
    }

    document.getElementById(id).innerHTML = out;

    /* scroll to bottom of spot div if its already at the bottom (tail -f behavior) */
    if (id == "spots" && at_bottom) {
        scrollToBottom(id);
    }
}

/*
* Scroll to the bottom of an overflow div with given ID
*/
function scrollToBottom(id) {
    var div = document.getElementById(id);
    var scrollHeight = Math.max(div.scrollHeight, div.clientHeight);
    div.scrollTop = scrollHeight - div.clientHeight;
}

/*
* Write the skimmer status data to the status div.
*
* param: jsonData: an array of text lines in a json array.
*/
function writeStatus(jsonData) {
    var i = 0;
    var out = '<pre>';

    for (i = 1; i < jsonData.length; i++) {
        out += '<div class="statusline">';
        out += jsonData[i];
        out += '</div>';
    }
    out += '</pre>';

    document.getElementById('status').innerHTML = out;
}

/**
 * Continually call getstatus endpoint and print it out until the app is running
 */
function get_status() {
    var q = new XMLHttpRequest();
    q.open("get", '/getstatus', true);
    q.send();

    q.onload = function () {
        var v = JSON.parse(q.responseText);

        writeStatus(v);

        // v[0] is the current cSkimmer state value. either starting or running
        if (v[0] != "running")
            setTimeout(get_status, 1000);

        if (v[0] === "running")
            $("#loadSpinner").hide();
        else if (v[0] === "starting")
            $("#loadSpinner").show();
    };
};

/**
 * Calls version endpoint and prints the version number to DOM
 */
function get_version_num() {
    var q = new XMLHttpRequest();
    q.open("get", '/version', true);
    q.send();

    q.onload = function () {
        var v = JSON.parse(q.responseText);
        console.log(v);
        $("#verNum").text("Web App version: " + v.version);
    };
};

/**
 * call the load endpoint to get the app's saved scratchpad
 */
function get_saved_scratchpad() {
    var q = new XMLHttpRequest();
    q.open("get", '/load', true);
    q.setRequestHeader("Content-Type", "application/json;charset=UTF-8");
    q.send();

    q.onload = function () {
        var v = JSON.parse(q.responseText);
        v.forEach(el => $('#scratchpad').append(getSpotHtml(el)));
    };
};

/**
 * Call save endpoint and pass in the spot html
 */
function save_spot(spot) {
    var q = new XMLHttpRequest();
    q.open("post", '/save/', true);
    q.setRequestHeader("Content-Type", "application/json;charset=UTF-8");
    q.send(JSON.stringify(spot));
};

/**
 * Continually set the UTC time clock on the page
 */
function updateClock() {
    var t = new Date().toUTCString();
    document.getElementById('time').innerHTML = t;

    // call this function again in 1000ms
    setTimeout(updateClock, 1000);
};


/*
*
* --- Main Script Area ---
*
*/

/* WEBSOCKET SETUP */
const spotsWebsocket = new WebSocket('ws://' + location.host + '/getspotsws');
spotsWebsocket.addEventListener('message', ev => {
    jsonData = JSON.parse(ev.data);
    writeSpotLines(jsonData, "spots");
    spots = jsonData;
});

const skedsWebsocket = new WebSocket('ws://' + location.host + '/getskedsws');
skedsWebsocket.addEventListener('message', ev => {
    jsonData = JSON.parse(ev.data);
    writeSpotLines(jsonData, "skeds");
    skeds = jsonData;
});


/* initial stuff */
get_status(); 
updateClock();
get_version_num();
get_saved_scratchpad();


/* toggle the Highlight on a spotline on click */
$(document).on("click", 'div.spotline', function (event) {

    var parentId = $(this).parents("div").attr('id');
    if (parentId === "scratchpad")
        return;

    if ($(this).hasClass("highlight"))
        $(this).removeClass("highlight");
    else {
        /* add to the scratch pad div */
        let savedLine = $(this).clone();
        $('#scratchpad').append(savedLine);

        /* highlight the line in-place */
        $(this).addClass("highlight");

        let cs = $(this).find('.callsign');
        let toFind = cs[0].innerHTML.trim();

        if ($(event.target).parents("#spots").length > 0) {
            var f = spots.find(el => el.call === toFind);
        }
        else {
            var f = skeds.find(el => el.call === toFind);
        }

        save_spot(f);
    }
});

/* Detect if the user has scrolled to bottom of SPOTS div */
$('#spots').on('scroll', function () {
    var x = this.scrollHeight - Math.round(this.scrollTop);

    if (x === this.clientHeight) {
        at_bottom = true;
    }
    else {
        /* user scrolled back up. this flag is now set so the
        spot list doesn't auto jump to bottom of div */
        at_bottom = false;
    }
});


/*
*
*  - Button handlers used in html files -
*
*/

/*
* clear the scratch pad div and clear the session data via http post
*/
function onScratchPadClearClick() {
    $('#scratchpad').empty();

    var q = new XMLHttpRequest();
    q.open("post", '/save/?clear=1', true);
    q.send();
}


/*
*  Play button click handler. 
*/
function onPlayClick() {
    scrollToBottom("spots");
}


/*
 * Restart button click handler
 */
function onRestartClick() {
    var q = new XMLHttpRequest();
    q.open("post", '/restart/', true);
    q.send();

    /* when post is done */
    q.onreadystatechange = () => {
        if (q.readyState === XMLHttpRequest.DONE) {
            const status = q.status;
            if (status === 0 || (status >= 200 && status < 400)) {
                /* kick off the status check ajax */
                get_status();
            }
        }
    };
}